src/AsyncBridge/AsyncCompatLibExtensions.cs

```diff
@@ -85,5 +97,231 @@ public static ConfiguredTaskAwaitable ConfigureAwait(this Task task, bool contin
+    /// not been canceled already.
+    /// </para>
+    /// <para>
+    /// Subsequent calls to CancelAfter will reset the delay for this  
+    /// <see cref="T:System.Threading.CancellationTokenSource"/>, if it has not been
+    /// canceled already.
+    /// </para>
+    /// </remarks>
+    public static void CancelAfter(this CancellationTokenSource cancelSource, TimeSpan delay)
+    {
+        if (cancelSource == null)
+            throw new ArgumentNullException(nameof(cancelSource));
+
+        if (delay < _timeoutInfinite)
+            throw new ArgumentOutOfRangeException(nameof(delay));
+
+        if (cancelSource.IsCancellationRequested)
+            return;
+
+        Timer MyTimer = null;
+
+        // CTS claims it's thread-safe for all methods, so we should be the same.
+        // If someone calls CancelAfter concurrently for the same CTS, we won't explode or cause a memory leak
+        do
+        {
+            if (_cancelTimers.TryGetTimer(cancelSource, out var CurrentTimer))
+            {
+                if (MyTimer != null)
+                    MyTimer.Dispose();
+
+                MyTimer = CurrentTimer;
+                break; // Reset the existing timer to the new delay
+            }
+
+            // An active timer doesn't hold a strong reference to itself, so adding the CTS as state won't hold it alive
+            MyTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
+
+        } while (!_cancelTimers.TryAddTimer(cancelSource, MyTimer));
+
+        try
+        {
+            MyTimer.Change(delay, _timeoutInfinite);
+        }
+        catch (ObjectDisposedException)
+        {
+        }
+    }
+
+    private static void OnCancelAfterTimer(object state)
+    {
+        var CancelSource = (CancellationTokenSource)state;
+
+        if (!_cancelTimers.TryRemoveTimer(CancelSource, out var MyTimer))
+            return;
+
+        MyTimer.Dispose();
+
+        try
+        {
+            CancelSource.Cancel();
+        }
+        catch (ObjectDisposedException) // If the cancellation token has been disposed of, ignore the exception
+        {
+        }
+    }
+
+#if NET35
+    private struct WeakCTS : IEquatable<WeakCTS>
+    {
+        private readonly int _hashCode;
+
+        internal WeakCTS(CancellationTokenSource cts)
+        {
+            _hashCode = cts.GetHashCode();
+            CTS = new WeakReference(cts);
+        }
+
+        public override int GetHashCode()
+        {
+            return _hashCode;
+        }
+
+        public override bool Equals(object obj)
+        {
+            return obj is WeakCTS Value && Equals(Value);
+        }
+
+        public bool Equals(WeakCTS other)
+        {
+            // In the case our CTS is no longer alive, this will still work when doing dictionary lookups. Null matches null
+            return _hashCode == other._hashCode && ReferenceEquals(CTS.Target, other.CTS.Target);
```

Comment:
Is there a reason for the `_hashCode` check here?

Reply:
Possibly not. I wouldn't expect the ConcurrentDictionary to call Equals on objects without matching GetHashCode values. But that would rely on an implementation detail. Having the check ensures it can't go wrong. Rackspace.Threading does it too.

Reply:
> I wouldn't expect the ConcurrentDictionary to call Equals on objects without matching GetHashCode values. 

Let's say it did. Either way, removing this check won't cause the wrong answer to be returned. Right? 
It must be an optimization.

Reply:
Without the check, Equals could return true if the WeakCTS you compare against has also been GC'ed (so it's WeakReference gives null too), even if it has a different hash. It'd then remove the wrong item from the ConcurrentDictionary, potentially skipping the item you intended to remove and leaving it sitting in the ConcurrentDictionary until the next cleanup.

Reply:
This is another incomplete solution then, since hash codes are expected to collide. Does it matter which null `other.CTS.Target` gets removed if `CTS.Target` is null?